### PR TITLE
Fix #492. Traits that require an outer pointer are not SAMs.

### LINKED ISSUE #492: Inner traits are sometimes thought to be SAM types when they aren't
The following code:
```scala
class Outer {
  trait Inner {
    def apply(x: Int): Int
  }
  val f: Inner = x => x
}
```
Becomes after frontend:
```scala
package <empty> {
  class Outer() extends Object() { 
    <trait> trait Inner() extends Object { 
      def apply(x: Int): Int
    }
    val f: Outer.this.Inner = {
      def $anonfun(x: Int): Int = x
      closure($anonfun:Outer.this.Inner)
    }
  }
```
So far so good: `Inner` is a SAM type (also known as a [Functional Interface](https://docs.oracle.com/javase/8/docs/api/java/lang/FunctionalInterface.html)) so it can be the result of a closure, now watch what happens after ExplicitOuter:
```scala
class Outer() extends Object() { 
  <trait> trait Inner() extends Object() { 
    def apply(x: Int): Int
    final def Outer$Inner$$$outer: Outer
  }
...
```
Oh no! Inner now has two abstract methods, it's no longer a SAM type, and that's terrible.

I'm not sure but I think that there are two things wrong here:
- The `SAMType` extractor should not consider traits that may get an outer pointer as SAM types, not sure what the best way to ensure that is.
- In this example, it seems like `Inner` should have the flag `PureInterface` set, that way it won't get an outer pointer. I investigated a bit: `NormalizeFlags` should set `PureInterface` if `NoInits` is set but that flag isn't set on `Inner` either.  `Namer#denotsNamed` should set `NoInits`, but it doesn't because `TreeInfo#isNoInitMember` returns false when called on the apply `DefDef`. That's where I stopped.

@odersky : Could you take a look at this and advise on the best way to fix it?

### PR DESCRIPTION
**Pull request by DarkDimius**

### CODE REVIEW

File: tests/pos/sammy_poly.scala [hunk 1]
```diff
@@ -1,7 +1,13 @@
 // test synthesizeSAMFunction where the sam type is not fully defined
+
+trait F1[T, U] { def apply(x: T): U }
 class T {
-  trait F[T, U] { def apply(x: T): U }
   // NOTE: the f(x) desugaring for now assumes the single abstract method is called 'apply'
-  def app[T, U](x: T)(f: F[T, U]): U = f(x)
-  app(1)(x => List(x))
+  def app1[T, U](x: T)(f: F1[T, U]): U = f(x)
+  def app2[T, U](x: T)(f: F2[T, U]): U = f(x)
```
Comment (smarter) [inline tests/pos/sammy_poly.scala]: Should be `T.F2` instead of `F2`, or add `import T._`

**Comment by DarkDimius** [comment]
@smarter please review. Does this unblock you?

**Comment by smarter** [comment]
It should yes, though I have a fix/workaround to avoid test failures on #411 I'll push soon.

**Comment by smarter** [comment]
LGTM but this does not prevent all the "fake" SAMs, as noted in https://github.com/lampepfl/dotty/issues/492#issuecomment-95376819
```scala
trait X {
  def foo(): Int = 3
}

trait FakeSAM extends X {
  def apply(x: Int): Int
  override def foo(): Int = super.foo()
  // Get extra abstract method "private def super$foo(): Int"
}
```
And
```scala
trait FakeSAM2 {
  def apply(x: Int): Int
  val dummy: Int = 3
  // the val becomes:
  // <accessor> def dummy(): Int <--- extra abstract method
  // protected def initial$dummy(): Int = 3
}
```

I don't actually require this to be merged for #411 as I can easily avoid the test failure, so don't feel forced to merge it now.

**Comment by DarkDimius** [comment]
I am aware of super-calls. But I do not want to make changes to handling of `super-accessors` before the refactoring is merged. I'll make it in a saparate PR later.

**Comment by DarkDimius** [comment]
Second example is not a SAM - SAMs cannot have fields.

**Comment by smarter** [comment]
That's not a requirement of https://docs.oracle.com/javase/8/docs/api/java/lang/FunctionalInterface.html, static fields are allowed.

**Comment by DarkDimius** [comment]
there's no such thing as `static field` in scala.
The only `static` thing that exists in scala is forwarders.

**Comment by odersky** [comment]
LGTM also

**Pull request merged**